### PR TITLE
Replace `embedded-consul` with Testcontainers Hashicorp Consul module

### LINKED ISSUE #4936: Use Testcontainers Hashicorp Consul Module instead of embedded-consul for testing
https://github.com/pszymczyk/embedded-consul has been deprecated in favour of [Testcontainers Hashicorp Consul Module](https://www.testcontainers.org/modules/consul/)

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

`embedded-consul` has been deprecated in favor of Testcontainers
Hachicorp Consul module according to its author: https://github.com/pszymczyk/embedded-consul

Modifications:

- Replaced `embedded-consul` with Testcontainers Hashicorp Consul module
  - Removed all dependencies on `embedded-consul`
  - Added the dependencies on Testcontainers
    - We're now using `ConsulContainer` to configure and start Consul.
  - Replaced `127.0.0.1` or `localhost` with `host.docker.internal`
    because Consul is now running inside the Docker container
- Fixed a bug where `aclConfiguration()` generates invalid JSON
- Added `@EnabledIfDockerIsAvailable` annotation so that the tests are
  run only when Docker is available.
- Miscellaneous:
  - Added `ConsulTestBase.consulUri()` to deduplicate URI creation

Result:

- We don't depend on a deprecated component anymore.
- Closes #4936

### CODE REVIEW

File: consul/src/test/java/com/linecorp/armeria/internal/consul/EnabledIfDockerIsAvailable.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import org.junit.jupiter.api.condition.EnabledIf;
+
+@Inherited
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+@EnabledIf(
+        value = "com.linecorp.armeria.internal.consul.ConsulTestBase#isDockerAvailable",
+        disabledReason = "Docker is not available."
+)
+public @interface EnabledIfDockerIsAvailable {
```
Comment (jrhee17) [inline consul/src/test/java/com/linecorp/armeria/internal/consul/EnabledIfDockerIsAvailable.java:33]: Just to make sure, is there a way to check if our CI has docker available?
If not, I guess we can just try merging and check gradle enterprise if the test is run

**Review by jrhee17 — APPROVED** [review]
Thanks for the cleanup @trustin 🙇 👍 🙇

File: dependencies.toml [hunk 2]
```diff
@@ -275,10 +276,6 @@ version.ref = "checkerframework"
 module = "com.puppycrawl.tools:checkstyle"
 version.ref = "checkstyle"
 
-[libraries.consul]
-module = "com.pszymczyk.consul:embedded-consul"
-version.ref = "consul"
```
Comment (minwoox) [inline dependencies.toml:280]: Could you also remove `consul` from the version section?

**Review by minwoox — APPROVED** [review]
One nit. Thanks! 👍

File: consul/src/test/java/com/linecorp/armeria/internal/consul/EnabledIfDockerIsAvailable.java (hunk 1 above)
Comment (ikhoon) [reply consul/src/test/java/com/linecorp/armeria/internal/consul/EnabledIfDockerIsAvailable.java:33]: Maybe not for self-hosted runners. Let me install it soon.

**Review by ikhoon — APPROVED** [review]
Thanks!

File: consul/src/test/java/com/linecorp/armeria/internal/consul/EnabledIfDockerIsAvailable.java (hunk 1 above)
Comment (ikhoon) [reply consul/src/test/java/com/linecorp/armeria/internal/consul/EnabledIfDockerIsAvailable.java:33]: I have just installed docker to our self-hosted runners.

**Comment by ikhoon** [comment]
I triggered CI. Let's see if Testcontainers correctly runs on self-hosted runners.

File: consul/src/test/java/com/linecorp/armeria/internal/consul/ConsulTestBase.java [hunk 3]
```diff
@@ -58,101 +55,63 @@
  * Selectively disable Consul tests to suppress the stressful flakiness.
  */
 @FlakyTest
+@Testcontainers
+@EnabledIfDockerIsAvailable
```
Comment (anuraaga) [inline consul/src/test/java/com/linecorp/armeria/internal/consul/ConsulTestBase.java:59]: I think `@Testcontainers` already has an option for this

https://github.com/open-telemetry/opentelemetry-java/blob/ad6cbfc4b49b1c996b588a3942d665034468da21/integration-tests/otlp/src/main/java/io/opentelemetry/integrationtest/OtlpExporterIntegrationTest.java#L106

Comment (trustin) [reply consul/src/test/java/com/linecorp/armeria/internal/consul/ConsulTestBase.java:59]: Ahhh, thanks for the heads up, @anuraaga :bow:

**Comment by ikhoon** [comment]
> I triggered CI. Let's see if Testcontainers correctly runs on self-hosted runners.

Tests with Testcontainers weren't activated although docker is available.  
```
[reports]$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., 0.0.0+unknown)
 ...

[reports]$ find tests/
tests/
tests/shadedTest
tests/shadedTest/index.html
tests/shadedTest/packages
tests/shadedTest/packages/com.linecorp.armeria.client.consul.html
tests/shadedTest/classes
tests/shadedTest/classes/com.linecorp.armeria.client.consul.ConsulEndpointGroupBuilderTest.html
tests/shadedTest/css
tests/shadedTest/css/base-style.css
tests/shadedTest/css/style.css
tests/shadedTest/js
tests/shadedTest/js/report.js
[reports]$ 
```

File: consul/src/test/java/com/linecorp/armeria/internal/consul/ConsulTestBase.java (hunk 3 above)
Comment (ikhoon) [reply consul/src/test/java/com/linecorp/armeria/internal/consul/ConsulTestBase.java:59]: I was able to see the test results in CI with `@Testcontainers(disabledWithoutDocker = true)` as @anuraaga said.
```
./tests/test/classes
./tests/test/classes/com.linecorp.armeria.client.consul.ConsulEndpointGroupBuilderTest.html
./tests/test/classes/com.linecorp.armeria.client.consul.ConsulEndpointGroupTest.html
./tests/test/classes/com.linecorp.armeria.internal.consul.ConsulClientBuilderTest.html
./tests/test/classes/com.linecorp.armeria.server.consul.ConsulUpdatingListenerTest.html
```

**Comment by trustin** [comment]
@ikhoon Could you check https://github.com/line/armeria/actions/runs/5726121065?pr=4970 once it's finished? It now uses `@Testcontainers(disabledWithoutDocker = true)`

**Comment by ikhoon** [comment]
> @ikhoon Could you check [line/armeria/actions/runs/5726121065?pr=4970](https://github.com/line/armeria/actions/runs/5726121065?pr=4970) once it's finished? It now uses `@Testcontainers(disabledWithoutDocker = true)`

I was able to run tests with Testcontainers by executing CLI commands on our CI machines but it wasn't executed with GitHub runners. I am troubleshooting the problem while testing possible workarounds such as https://github.com/actions/runner/issues/411

**Comment by ikhoon** [comment]
I realized that the Consul tests are marked as `@FlakyTest` but I check the test results on the normal CI machines. 😅

**Comment by ikhoon** [comment]
I confirmed that all tests correctly ran with Testcontainers on self-hosted runners.

**Pull request merged**